lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -1471,10 +1471,19 @@ LogicalResult InferenceMapping::mapOperation(Operation *op) {
         constrainTypes(sel, solver.known(1));
         maximumOfTypes(op.getResult(), op.getHigh(), op.getLow());
       })
-
-      // Handle the various connect statements that imply a type constraint.
-      .Case<FConnectLike>(
+      .Case<UninferredWidthCastOp>([&](auto op) {
+        declareVars(op.getResult(), op.getLoc());
+        constrainTypes(op.getResult(), op.getInput());
+      })
+      .Case<ConnectOp>(
           [&](auto op) { constrainTypes(op.getDest(), op.getSrc()); })
+      .Case<RefDefineOp>(
+          [&](auto op) { constrainTypes(op.getDest(), op.getSrc()); })
+      // StrictConnect is an identify constraint
+      .Case<StrictConnectOp>([&](auto op) {
```

Comment:
This very likely is not going to work as expected, been looking into this for ensuring inferred widths are identical across reference "networks" (and also don't back-propagate / participate in minimum-width-that-satisfies-constraints behavior, which strictconnect probably can/does want to).

I think if we add a equality constraint explicitly that'll get us what we want.

As-is this will propagate both directions causing rather different and perhaps unexpected behavior.
Not sure about this, but poking on something similar I managed to get this to change widths of circuit components that were statically known.  Whoops.

Anyway, compare how this design is after InferWidths on latest main vs your PR to see what I mean:

```
circuit Width:
  module SubWidth:
    input in : UInt
  module Width:
    wire x : UInt
    
    inst s of SubWidth
    s.in <= x

    x <= UInt<5>(0)


    node z = UInt<10>(5)
    inst u of SubWidth
    u.in <= z
```
(is `x` 5 or 10?)

This is a similar example that's currently rejected but this accepts:

```
circuit Width:
  module SubWidth:
    input in : UInt
  module Width:
    wire x : UInt
    wire y : UInt
    
    inst s of SubWidth
    s.in <= x

    inst t of SubWidth
    t.in <= y

    y is invalid
    x <= UInt<5>(0)


    node z = UInt<10>(5)
    inst u of SubWidth
    u.in <= z
```

Reply:
It is what was happening before.

Reply:
In this example, the ports are the same, but instead of:
`const<5> -> wire<5> -> implicit extend<5-10> -> instance<10>`
we get:
`const<10> -> wire<10> -> instance<10>`

Which is arguably better in terms of less implicit ops.

Reply:
unify seems to override existing constraints, which is bad (tm)

Reply:
Ooh, this was subtle.  This code was fine, but emitConnect wasn't.

Reply:
```// -----// IR Dump After InferWidths (firrtl-infer-widths) //----- //
firrtl.circuit "Width" {
  firrtl.module private @SubWidth(in %in: !firrtl.uint<10>) {
  }
  firrtl.module @Width() attributes {convention = #firrtl<convention scalarized>} {
    %x = firrtl.wire : !firrtl.uint<5>
    %s_in = firrtl.instance s @SubWidth(in in: !firrtl.uint<10>)
    %0 = firrtl.pad %x, 10 : (!firrtl.uint<5>) -> !firrtl.uint<10>
    firrtl.strictconnect %s_in, %0 : !firrtl.uint<10>
    %c0_ui5 = firrtl.constant 0 : !firrtl.uint<5>
    firrtl.strictconnect %x, %c0_ui5 : !firrtl.uint<5>
    %c5_ui10 = firrtl.constant 5 : !firrtl.uint<10>
    %z = firrtl.node %c5_ui10 : !firrtl.uint<10>
    %u_in = firrtl.instance u @SubWidth(in in: !firrtl.uint<10>)
    firrtl.strictconnect %u_in, %z : !firrtl.uint<10>
  }
}```

Reply:
So now `const<5> -> wire<5> -> explicit extend<5-10> -> instance<10>`